cupy/core/_fusion_analysis.py

```diff
@@ -0,0 +1,612 @@
+
+        # Parse inputs.
+        in_param = self._from_arraylike_interface(a)
+
+        if not isinstance(in_param, _FusionCudaArray):
+            raise NotImplementedError(
+                'Reduction for scalar arguments is not supported.')
+
+        axes = _normalize_axis(axis, in_param.ndim)
+
+        if dtype is not None:
+            dtype = numpy.dtype(dtype)
+
+        if keepdims:
+            raise NotImplementedError('keepdims is not supported.')
+
+        # Determine the shape of out_param.
+        out_ashape = tuple([
+            d for axis, d in enumerate(in_param.ashape) if axis not in axes])
+        out_rshape = tuple([
+            d for axis, d in enumerate(in_param.rshape) if axis not in axes])
+
+        # Rotate axes.
+        # This condition is only for performance improvement,
+        if not all([i == axis for i, axis in enumerate(axes)]):
+            in_param = self.vc.rotate_with_axis(in_param, axes)
+
+        # Get operation code from dtypes.
+        in_dtypes, out_dtypes, op_dtype, reduce_op = _get_reduction_dtypes_op(
+            reduce_func, in_param, dtype)
+        if out is None:
+            out_param = self.vc.generate_new_array(
+                op_dtype, out_rshape, out_ashape)
+        else:
+            out_param = self._from_arraylike_interface(out)
+            if out_param.rshape != out_rshape:
+                raise ValueError(
+                    'Shape of specified output variable is not consistent '
+                    'with reduced shape.')
+
+        # Make submodule.
+        name = 'reduce{}'.format(len(self.op_list))
+        subm = _fusion_device_func._SubmoduleReduction(
+            name, reduce_func, reduce_op, [in_param], [out_param])
+        self.submodules[subm.name] = subm
+
+        # Register Op.
+        op = _fusion_op._FusionReductionOp(subm, in_param, out_param, axes)
+        self.op_list.append(op)
+
+        # Returns.
+        return self._to_interface(out_param)
+
+    def call_indexing(self, in_param, index):
+        """Call indexing routines.
+        """
+        in_param = self._from_arraylike_interface(in_param)
+        assert isinstance(in_param, _FusionCudaArray)
+        if isinstance(index, int):
+            out_param = self.vc.index_with_int(in_param, index)
+        elif isinstance(index, tuple):
+            out_param = self.vc.index_with_tuple(in_param, index)
+        else:
+            raise NotImplementedError
+
+        return self._to_interface(out_param)
+
+    def _call_function(self, func, args):
+        # Registers input variables.
+        in_params = []
+        for input_order, arg in enumerate(args):
+            if arg is None:
+                var = None
+            elif isinstance(arg, core.ndarray):
+                same_array_indices = [
+                    i for i in range(input_order) if arg is args[i]]
+                may_share_memory_indices = [
```

Comment:
`may_share_memory_indices` is not used if `len(same_array_indices) > 0`.
How about delaying this line?